core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1688,6 +1690,108 @@ RexNode simplifyAnd2ForUnknownAsFalse(List<RexNode> terms,
       return e;
     }
   }
+  
+  public RexNode simplifyBoolean(RexNode call) {
```

Comment:
Can you clarify how this differs from existing `simplifyOr...` and `simplifyAnd..` methods?

Reply:
simplifyAnd/Or handle the case like that "(a=b) and (7=8) => false" , it doesn't eliminates / extracts common factors.

Reply:
Common factors are eliminated by `org.apache.calcite.rex.RexUtil#flattenAnd` which is used in `org.apache.calcite.rex.RexUtil#composeConjunction(org.apache.calcite.rex.RexBuilder, java.lang.Iterable<? extends org.apache.calcite.rex.RexNode>, boolean)`.

Please make it explicit which patterns you are going to optimize, and why it requires new code rather than fixing the existing methods.
I see you've created CALCITE-4385, however, the issue description is blank, and the issue title is too generic: `Adding optimization to simplify the And/Or condition`.

The issue title and description should make it clear for the reviewer on what is the nature of the change, and why is it needed.

Reply:
Hi @vlsi The desc is added, and more details you can see in CALCITE_4375

Reply:
@vlsi RexUtil#composeConjunction just converts a collection of expressions into an AND, and do some true/false optimize.
```
  /**
   * Converts a collection of expressions into an AND.
   * If there are zero expressions, returns TRUE.
   * If there is one expression, returns just that expression.
   * If any of the expressions are FALSE, returns FALSE.
   * Removes expressions that always evaluate to TRUE.
   * Returns null only if {@code nullOnEmpty} and expression is TRUE.
   */
```

Reply:
As I said, flattendAnd and composeConjunctions remove duplicates: https://github.com/apache/calcite/blob/d03ce47536ddb2aaec6d445b3ea48464081a7ed0/core/src/main/java/org/apache/calcite/rex/RexUtil.java#L1234

Reply:
This set only de-dup case like: "a AND a".
In my case, the common part is in "OR", the method doesn't split "OR" then it will not "addAnd":  "(a=b || c=7) && (a=b || c=7 || d=8)".

Reply:
And we are not only eliminating the common part, but we also need to extract it and re-construct the condition.

Reply:
Can we resolve this part?